RecoLocalCalo/EcalRecAlgos/src/KernelHelpers.h

```diff
@@ -9,6 +13,557 @@ uint32_t hashedIndexEB(uint32_t id);
+                for (int k=0; k<j; ++k)
+                    sumsq2 += L(i, k) * L(j, k);
+
+                // comput the i,j : i>j, elements to the left of the diagonal
+                auto const value_i_j = (m_i_j - sumsq2) / L(j, j);
+                L(i, j) = value_i_j;
+
+                // needed to compute diagonal element
+                sumsq += value_i_j * value_i_j;
+
+                total -= value_i_j * intermediate(j);
+            }
+
+            // second, compute the diagonal element
+            auto const l_i_i = std::sqrt(M(real_i, real_i) - sumsq);
+            L(i, i) = l_i_i;
+
+            intermediate(i) = total / l_i_i;
+        }
+    }
+};
+
+template<typename T, int N>
+struct FusedCholeskyForwardSubstUnrolled {
+    template<typename M1, typename M2, typename M3, typename M4>
+    __forceinline__ __device__ 
+    static void compute(
+            M1 const& M, 
+            M2 const& b,
+            M3 &L,
+            M4 &intermediate,
+            MapV<char> const& mapping) {
+        // compute element 0,0 for L
+        auto const real_0 = mapping(0);
+        auto const sqrtm_0_0 = std::sqrt(M(real_0, real_0));
+        L(0, 0) = sqrtm_0_0;
+
+        // compute solution for forward subst for element 0
+        auto const interm_0 = b(real_0) / sqrtm_0_0;
+        intermediate(0) = interm_0;
+
+        #pragma unroll
+        for (int i=1; i<N; ++i) {
+            // load the value to sub from
+            auto const real_i = mapping(i);
+            T total = b(real_i);
+
+            // first compute elements to the left of the diagoanl
+            T sumsq{static_cast<T>(0)};
+            for (int j=0; j<i; ++j) {
+                T sumsq2{static_cast<T>(0)};
+                auto const real_j = mapping(j);
+                auto const m_i_j = M(real_i, real_j);
+                for (int k=0; k<j; ++k)
+                    sumsq2 += L(i, k) * L(j, k);
+
+                // comput the i,j : i>j, elements to the left of the diagonal
+                auto const value_i_j = (m_i_j - sumsq2) / L(j, j);
+                L(i, j) = value_i_j;
+
+                // needed to compute diagonal element
+                sumsq += value_i_j * value_i_j;
+
+                total -= value_i_j * intermediate(j);
+            }
+
+            // second, compute the diagonal element
+            auto const l_i_i = std::sqrt(M(real_i, real_i) - sumsq);
+            L(i, i) = l_i_i;
+
+            intermediate(i) = total / l_i_i;
+        }
+    }
+};
+
 }}
 
+__forceinline__ __device__
+bool cudaCheckFromDevice_(
+        char const* file, int const line, 
+        char const* cmd, cudaError_t result) {
+    if (result == cudaSuccess)
+        return true;
+
+    char const* error = cudaGetErrorName(result);
+    char const* msg = cudaGetErrorString(result);
+    printf("\n%s, line %d:\ncudaCheckFromDevice(%s);\n%s: %s\n",
+        file, line, cmd, error, msg);
+    assert(0);
```

Comment:
Just to note that we have experienced an `assert()` in kernel code (or in a function called by a kernel) leading to a device-side synchronization between CUDA streams, see #186 and #183. On the other hand I see I had noted in https://github.com/cms-patatrack/cmssw/issues/183#issuecomment-427492665 that in case the condition can be checked at compile time the `assert` would not lead to syncrhonization.

Reply:
thx a lot Mattti! I did not know about this at all! just to note, i added this to provide at least some check on launching kernels from device side. Feel free to recommend a better check in here... this was copied from what you guys have in cudaCheck...

Reply:
I don't really have better ideas. But check with the profiler whether the kernels get run in parallel or not , the effect should be rather visible (see discussion and profile timelines in #183).